### PR TITLE
Softmax backward & Cross Entropy backward should both be optimal now

### PR DESCRIPTION
**Pull request by fufjvnvnf**
Resolves #7

### CODE REVIEW

**Comment by fufjvnvnf** [comment]
Resolves #7

**Comment by stevenygd** [comment]
I wouldn't say it resolves #7 . Since the benchmark shows that the improvement is quite marginal:

# Before:
```
Old forward uses on average 0.008957189829999998 seconds
New forward uses on average 0.010831998020000002 seconds
Old backward uses on average 0.005899088019999998 seconds
New backward uses on average 0.013368990589999996 seconds
Old forward allocates on average 4.24580416e6 bytes
New forward allocates on average 8.3952e6 bytes
Old backward allocates on average 2.10334016e6 bytes
New backward allocates on average 8.3952e6 bytes
```

# After
```
Old forward uses on average 0.009047210220000003 seconds
New forward uses on average 0.011174593820000002 seconds
Old backward uses on average 0.00592484136 seconds
New backward uses on average 0.013712984500000004 seconds
Old forward allocates on average 4.24579888e6 bytes
New forward allocates on average 8.3952e6 bytes
Old backward allocates on average 2.10333488e6 bytes
New backward allocates on average 8.3952e6 bytes
```

**Comment by fufjvnvnf** [comment]
Both backwards are now only one line with no loop. I will later refactor the original `SoftmaxCrossEntropyLoss` layer, whose speed is likely to decrease under the new float matrix interface. When we compare the benchmark of the two layers under the same interface we should have a clearer expectation for the speed of these layers.

**Comment by stevenygd** [comment]
@fufjvnvnf I'm not talking about the comparison between new layers and the old ones. The performance of the new layers stay the same before and after this PR.

**Comment by fufjvnvnf** [comment]
Closing this PR but keeping the issue open. I'll also keep these changes in my next PR which modifies the interface of `SoftmaxCrossEntropy` layer since it still increases the speed slightly and the code looks cleaner & nicer.
